src/client/render/plain.cpp

```diff
@@ -19,58 +19,141 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 #include "plain.h"
-#include "settings.h"
+#include "secondstage.h"
+#include "client/camera.h"
+#include "client/client.h"
+#include "client/clientmap.h"
+#include "client/hud.h"
+#include "client/minimap.h"
+#include "client/shadows/dynamicshadowsrender.h"
 
-inline u32 scaledown(u32 coef, u32 size)
+/// Draw3D pipeline step
+void Draw3D::run(PipelineContext *context)
 {
-	return (size + coef - 1) / coef;
+	if (m_target)
+		m_target->activate(context);
+
+	context->device->getSceneManager()->drawAll();
+	context->device->getVideoDriver()->setTransform(video::ETS_WORLD, core::IdentityMatrix);
+	if (!context->show_hud)
+		return;
+	context->hud->drawBlockBounds();
+	context->hud->drawSelectionMesh();
+	if (context->draw_wield_tool)
+		context->client->getCamera()->drawWieldedTool();
 }
 
-RenderingCorePlain::RenderingCorePlain(
-	IrrlichtDevice *_device, Client *_client, Hud *_hud)
-	: RenderingCore(_device, _client, _hud)
+void DrawHUD::run(PipelineContext *context)
 {
-	scale = g_settings->getU16("undersampling");
+	if (context->show_hud) {
+		if (context->shadow_renderer)
+			context->shadow_renderer->drawDebug();
+
+		if (context->draw_crosshair)
+			context->hud->drawCrosshair();
+
+		context->hud->drawHotbar(context->client->getEnv().getLocalPlayer()->getWieldIndex());
+		context->hud->drawLuaElements(context->client->getCamera()->getOffset());
+		context->client->getCamera()->drawNametags();
+		auto mapper = context->client->getMinimap();
+		if (mapper && context->show_minimap)
+			mapper->drawMinimap();
+	}
+	context->device->getGUIEnvironment()->drawAll();
 }
 
-void RenderingCorePlain::initTextures()
+
+void MapPostFxStep::setRenderTarget(RenderTarget * _target)
 {
-	if (scale <= 1)
-		return;
-	v2u32 size{scaledown(scale, screensize.X), scaledown(scale, screensize.Y)};
-	lowres = driver->addRenderTargetTexture(
-			size, "render_lowres", video::ECF_A8R8G8B8);
+	target = _target;
 }
 
-void RenderingCorePlain::clearTextures()
+void MapPostFxStep::run(PipelineContext *context)
 {
-	if (scale <= 1)
-		return;
-	driver->removeTexture(lowres);
+	if (target)
+		target->activate(context);
+
+	context->client->getEnv().getClientMap().renderPostFx(context->client->getCamera()->getCameraMode());
 }
 
-void RenderingCorePlain::beforeDraw()
+void RenderShadowMapStep::run(PipelineContext *context)
 {
-	if (scale <= 1)
-		return;
-	driver->setRenderTarget(lowres, true, true, skycolor);
+	// This is necessary to render shadows for animations correctly
+	context->device->getSceneManager()->getRootSceneNode()->OnAnimate(context->device->getTimer()->getTime());
```

Comment:
is care taken that this isn't called twice per frame in that case?

Reply:
This is old code, and we can't really do a lot here without rewriting ShadowRenderer.